### PR TITLE
[WIP] Documentation FAQ & General Improvements

### PR DESCRIPTION
**Pull request by Still34**
# Preview

[Build status](https://ci.appveyor.com/project/Still/discord-net)

The latest preview can be seen [here](https://docs.stillu.cc). 

It has been set up with AppVeyor to be in sync with the commits from the PR branch.

# Purpose

This PR aims to do the following,

1. Improve the XMLDocs in many sections of the library. This includes classes, props, methods, etc.
2. Add a FAQ section under the current documentation page.
3. Improve the general documentations.
4. General cleanup (add `ConfigureAwait(false)`; remove unused references) in the codebase.

This PR is currently under work-in-progress. Any suggestions are welcome.

# To-do

- [ ] Finish covering common entities' XMLDocs
    - This includes at least the base interfaces and classes.
- [ ] Finish building FAQ
    - [ ] Add "I want to build a bot with a UI" section.
    - [ ] Add "Entity Comparison" section.
- [x] Proofread guide to ensure that it is up-to-date
- [x] Enable `View Source` button
    - Added in 6f4a4e0.
    * Current resolution in the commit is not ideal, please check the commit detail to see more regarding this issue.
- [x] Find a workaround for constantly changing method signature names causing xref to fail
    - Resolved in 291ad1cf. 
    - Method names will require a wildcard symbol at the end of the full name to redirect to the proper method. e.g. ``Discord.Commands.ModuleBase`1.ReplyAsync*``
    - Event and class names do not require any wildcard symbol.
- [x] Enable site-wide search via docfx
   - Resolved in 68be9d3.
   - Searching in documentation for a method was often difficult, or had to rely on separate service such as Googling the host (e.g. `SendMessageAsync site:discord.foxbot.me`) or .NET Discord bot. The built-in search function for docfx should help aid this.
   - This was enabled via `_enableSearch` under `docfx.json`.
- [x] Fix foreign cross-referencing
   - Resolved in 8be767b6.
   - An external cross-referencing service was required for this fix. Our `docfx.json` now references MSDN xref service.
   - This means objects that used to render `@System.IServiceProvider` now renders into `[IServiceProvider](https://docs.microsoft.com/dotnet/api/system.iserviceprovider)` automatically.
- [x] Create a dark template
   - Initial support added in 83435aa.
   - Theme switcher added in 6511b1a.

### CODE REVIEW

File: docs/faq/Commands.md [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+you want it to. Note that you should also learn Microsoft's 
+implementation of [Dependency Injection] before proceeding, as well 
+as how it works in [Discord.NET](../guides/commands/commands.md#usage-in-modules).
+
+A brief example of service and dependency injection can be seen below.
+
+[!code-csharp[DI](samples/commands/DI.cs)]
+
+[Dependency Injection]: https://docs.microsoft.com/en-us/aspnet/core/fundamentals/dependency-injection
+
```
Comment (Joe4evr) [inline docs/faq/Commands.md]: Consider adding a link to [this video](https://www.youtube.com/watch?v=QtDTfn8YxXg) as well.

File: docs/faq/Commands.md [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+## How does `RunMode.Async` work, and why is Discord.NET *not* using it by default?
+
+`RunMode.Async` works by spawning a new `Task` with an unawaited 
+[Task.Run], essentially making `ExecuteAsyncInternalAsync`, the task 
+that is used to invoke the command task, to be finished on a 
+different thread. This means that [ExecuteAsync] will be forced to
+return a successful [ExecuteResult] regardless of the execution.
+
+The following are the known caveats with `RunMode.Async`,
+	1. You introduce race condition.
```
Comment (Joe4evr) [inline docs/faq/Commands.md]: Race conditions should be plural here. The phrase a a whole can also use the word "potential", but I don't have a good one atm.

File: docs/faq/basic-operations.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+You may check the message channel type. Visit [Glossary] to see the 
+various types of channels.
+
+[Glossary]: Glossary.md#message-channels
+
+## How can I get the guild from a message?
+
+There are 2 ways to do this. You can do either of the following,
+	1. Cast the user as an [IGuildUser] and use its [IGuild] property.
+	2. Cast the channel as an [ITextChannel]/[IVoiceChannel] and use
```
Comment (Joe4evr) [inline docs/faq/basic-operations.md]: Should be "Cast as an `IGuildChannel`".

File: docs/faq/basic-operations.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+[!code-csharp[Casting](samples/basics/cast.cs)]
+
+[InvalidCastException]: https://docs.microsoft.com/en-us/dotnet/api/system.invalidcastexception
+[this post]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/types/how-to-safely-cast-by-using-as-and-is-operators
+
+## How do I send a message?
+
+Any implementation of [IMessageChannel] has a [SendMessageAsync]
+method. You can get the channel via [GetChannel] under the client.
+Remember, when using Discord.NET, polymorphism is a common recurring
```
Comment (Joe4evr) [inline docs/faq/basic-operations.md]: Should have a note that the returned channel object could also be an `IVoiceChannel`, which would be easier to understand that text messages can't be sent to such a channel.

File: docs/faq/basic-operations.md [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+[!code-csharp[Emoji](samples/basics/emoji.cs)]
+
+[AddReactionAsync]: xref:Discord.IUserMessage#Discord_IUserMessage_AddReactionAsync_Discord_IEmote_Discord_RequestOptions_
+  
+## Why am I getting so many preemptive rate limits when I try to add more than one reactions?
+
+This is due to how .NET parses the HTML header, mistreating 
+0.25sec/action to 1sec. This casues the lib to throw preemptive rate 
+limit more frequently than it should for methods such as adding 
+reactions.
```
Comment (Joe4evr) [inline docs/faq/basic-operations.md]: Is this really a .NET-specific bug? I thought it was an HTTP bug that header values can't be fractions, thus affecting *every* lib (and the others just get around it by implementing a special case).

File: docs/faq/getting-started.md [hunk 6]
```diff
@@ -0,0 +1,73 @@
+# Basic Concepts / Getting Started
+
+## How do I get started?
+First of all, welcome! Before you delve into using the library; 
+however, you should have some decent understanding of the language 
+you are about to use. This library touches on 
+[Task-based Asynchronous Pattern], [polymorphism], [interface] and
```
Comment (Joe4evr) [inline docs/faq/getting-started.md]: Could add `(TAP)` to show off the abbreviation.

File: docs/faq/getting-started.md [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+based on various conditions.
+![Snowflake Generation](images/snowflake.png)
+The ID can be seen by anyone; it is public. It is merely used to 
+identify an object in the Discord ecosystem. Many things in the 
+library require an ID to retrieve the said object. 
+  
+There are 2 ways to obtain the said ID. 
+  1. Enable Discord's developer mode. With developer mode enabled, 
+  you can - as an example - right click on a guild and copy the guild 
+  id (please note that this does not apply to Role IDs, see below).
```
Comment (Joe4evr) [inline docs/faq/getting-started.md]: This method also doesn't work for grabbing DM channel IDs in the client (you'll only get the target user's ID, which is *not the same ID*).

File: docs/faq/samples/commands/Remainder.cs [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+//		Coffee Cake
+[Command("echo")]
+public Task EchoRemainderAsync([Remainder]string text) => ReplyAsync(text);  
+
+//	Output:
+//		CommandError.BadArgCount
+[Command("echo-hassle")]
+public Task EchoAsync(string text) => ReplyAsync(text);
+
+//	The message would be seen as having 5 parameters, while the method
```
Comment (Joe4evr) [inline docs/faq/samples/commands/Remainder.cs]: *Five* parameters?

File: docs/guides/commands/commands.md [hunk 9]
```diff
@@ -153,15 +153,15 @@ Invoke [CommandService.AddModulesAsync] to discover modules and
...
 #### Loading Modules Manually
 
 To manually load a module, invoke [CommandService.AddModuleAsync] by
 passing in the generic type of your module and optionally, a
 dependency map.
 
-[CommandService.AddModuleAsync]: xref:Discord.Commands.CommandService#Discord_Commands_CommandService_AddModuleAsync__1
+[CommandService.AddModuleAsync]: xref:Discord.Commands.CommandService#Discord_Commands_CommandService_AddModuleAsync__1_System_IServiceProvider_
 
 ### Module Constructors
```
Comment (Joe4evr) [inline docs/guides/commands/commands.md]: I can't add a comment below here, but `Alternatively, you may accept an
IServiceProvider as an argument and extract services yourself.` should be discouraged.

File: src/Discord.Net.Commands/CommandError.cs [hunk 10]
```diff
@@ -1,26 +1,50 @@
...
+        /// </summary>
         UnknownCommand = 1,
 
         //Parse
+        /// <summary>
+        /// Thrown when the command fails to be parsed.
+        /// </summary>
         ParseFailed,
+        /// <summary>
+        /// Thrown when the input text has too few parameters.
```
Comment (Joe4evr) [inline src/Discord.Net.Commands/CommandError.cs]: "too few *or too many*" and input is *arguments*, not parameters. :upside_down_face:

File: src/Discord.Net.Commands/ModuleBase.cs [hunk 11]
```diff
@@ -21,11 +21,16 @@ public abstract class ModuleBase<T> : IModuleBase
         {
             return await Context.Channel.SendMessageAsync(message, isTTS, embed, options).ConfigureAwait(false);
         }
-
+        /// <summary>
+        /// The method to execute before executing the command.
+        /// </summary>
```
Comment (Joe4evr) [inline src/Discord.Net.Commands/ModuleBase.cs]: Missing the `<param>`.

**Review by Joe4evr — COMMENTED** [review]
Things I noticed.

File: docs/faq/basic-operations.md (hunk 5 above)
Comment (Still34) [reply docs/faq/basic-operations.md]: Fixed in edb64bde.

File: docs/faq/basic-operations.md (hunk 4 above)
Comment (Still34) [reply docs/faq/basic-operations.md]: Added in edb64bd.

File: docs/guides/commands/commands.md (hunk 9 above)
Comment (Still34) [reply docs/guides/commands/commands.md]: Added in edb64bd.

Comment (Still34) [reply docs/guides/commands/commands.md]: Added in edb64bd.

File: docs/faq/Commands.md (hunk 1 above)
Comment (Still34) [reply docs/faq/Commands.md]: Added in 179cd43a.

**Comment by Still34** [comment]
@foxbot what should we do about missing `TargetFramework` for `Discord.Net.Commands` and `Discord.Net.Webhook`? Ideally those should target NS1.3 as well for them to compile properly during docfx metadata compilation.

File: src/Discord.Net.Commands/Attributes/PreconditionAttribute.cs [hunk 12]
```diff
@@ -1,16 +1,22 @@
...
-        /// Specify a group that this precondition belongs to. Preconditions of the same group require only one
-        /// of the preconditions to pass in order to be successful (A || B). Specifying <see cref="Group"/> = <see langword="null"/>
-        /// or not at all will require *all* preconditions to pass, just like normal (A &amp;&amp; B).
+        ///     Specify a group that this precondition belongs to.
         /// </summary>
+        /// <remarks>
+        ///     <para>
+        ///          <see cref="Preconditions" /> of the same group require only one of the preconditions to pass in
+        ///          order to be successful (A || B). Specifying <see cref="Group" /> = <see langword="null" /> or not
+        ///          at all will require *all* preconditions to pass, just like normal (A && B).
```
Comment (Joe4evr) [inline src/Discord.Net.Commands/Attributes/PreconditionAttribute.cs]: Revert that `&&` back to `&amp;&amp;`.

**Comment by Still34** [comment]
Just FYI, this is currently being worked on on a separate branch `docs/faq-n-patches-offline`, and it will be squash merged occasionally to prevent spam on the PR branch.

**Comment by Still34** [comment]
Closed and will be re-opened when the PR is fully complete, as the current PR branch is obsolete anyways.

**Pull request closed without merging**